configuration-versioning/README.md

```diff
@@ -0,0 +1,137 @@
+```
+
+```python
+class ImageBuilderMetadataV2(ImageBuilderMetadataSchema):
+    image_distribution = fields.String(missing=None)
+    image_type: fields.String(missing=None)
+    image_architecture = fields.String(missing=None)
+    image_account_id = fields.String(missing=None)
+    packages_to_install = fields.String(missing=None)
+    owner = fields.String(missing=None)
+    project = fields.String(missing=None)
+```
+
+and the config object would have all the fields and the code could handle both
+
+- from the coding perspective this would not be a problem - the version would be detected either from whole config
+  or directly from the `image_builder_metadata`
+- but this may be confusing to users - versioning some configuration field apart from the whole config,
+  on the other hand it would be independent on the rest of the package config
+
+## Implementation details
+
+### Versioning
+
+- for each backwards incompatible change, we would bump the version
+- version would default to 1 (if not present)
+- this would be properly documented, we can get inspired by Docker,
+  [see](https://docs.docker.com/compose/compose-file/compose-versioning/#versioning), the documentation
+  would also include Packit version that starts supporting the particular configuration schema
+
+### Having more schemas vs having one schema
+
+#### Having more schemas
+
+- maintaining more versions of schema for whole package config should not be needed
+  - the schemas would be mostly have the same content, only with difference in few fields
+- this approach could be used for smaller parts of the configuration, e.g.
+  creating nested schema (as described for Example 1)
+- the coding around would be pretty straightforward, detect the version
+  and use the correct schema for loading/dumping accordingly
+
+#### Having one schema
+
+- in case of Example 2, the schema itself doesn't change
+- we could use the `pre_load` and `pre_dump` (we already use these) to migrate the data:
+
+```python
+    @staticmethod
+    def propagate_options_to_jobs(data, **_):
+        """
+        add all the fields (except for jobs) to every job so we can process only jobconfig in p-s
+        """
+        if not data:  # data is None when .packit.yaml is empty
+            return data
+        for job in data.get("jobs", []):
+            for k, v in data.items():
+                if k == "jobs":
+                    # overriding jobs doesn't make any sense
+                    continue
+                if self.version == 1:
+                    job.setdefault(k, v)
+                elif self.version == 2:
+                     # merge the values from top-level and the particular job
+                ...
+
+        return data
+```
+
+- we would need to make sure that load-dump-load results in same representation
+
+## Plan
+
+- decide whether/to which scope we want to do this in relation to [this issue](https://github.com/packit/research/issues/159)
```

Comment:
I feel like that as a part of #159 we will be able to do mainly simple changes (such as renaming). There may, however, be more complex changes to be made which we can't do automatically, as you mention later. I see one slight conflict of this with versioning. Imagine the following scenario:

1) user has a config of version `1`
2) we make a backwards incompatible change which cannot be handled automatically for users using this feature, they need to update manually to schema version `2`. But let's suppose that the user doesn't do this
3) We rename a field in the config, resulting in schema version `3`. We would be able to create a PR for this but we can't. Going version `1` -> `3` could break things because we can't "migrate" to changes done in `2`.

Reply:
That shouldn't really be a problem to implement but it will be another requirement for the automatic updates to take this into consideration

Reply:
yes, you are right, I just wanted to mention this in the research since it is somehow related

Reply:
Another option is to drop #159 once versioning is implemented.